File: onChangeMonitor.cpp
#include <iostream>
#include <filesystem>
#include <thread>
#include <chrono>
#include <map>
#include <vector>
#include <ctime>
#include <string>
#include "binning.h"
#include "multi_threaded_preprocessor.h"

namespace fs = std::filesystem;

// Function to scan a directory and return a map of file names and their modification times
std::map<std::string, fs::file_time_type> scan_directory(const std::string& directory) {
    std::map<std::string, fs::file_time_type> files;

    // Iterate over the contents of the directory
    for (const auto& entry : fs::directory_iterator(directory)) {
        if (fs::is_regular_file(entry.status())) {
            // Store the file name and its last modification time in the map
            files[entry.path().filename().string()] = fs::last_write_time(entry);
        }
    }

    return files;
}

// this parameter presents the lib length of 140
int lib_length = 140;

// path to the sequencing output - can be .txt or a .fastq file or a path to a folder with many .fastq files
// this is the file the includes all the basecalled reads that we want to processed
std::string design_file_path = "C:\\Users\\booki\\Desktop\\FunMoney\\DNA Storage\\concise_cpp\\deep_design.csv";

void monitor_directory(const std::string& directory, int interval = 5) {
    std::cout << "Monitoring directory: " << directory << std::endl;

    auto previous_files = scan_directory(directory);  // Initial snapshot of the directory

    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(interval));  // Wait for the specified interval

        auto current_files = scan_directory(directory);  // Scan the directory again

        // Check for new files
        std::vector<std::string> new_files;
        for (const auto& [file, _] : current_files) {
            if (previous_files.find(file) == previous_files.end()) {
                new_files.push_back(file);
            }
        }
        if (!new_files.empty()) {
            for (const auto& file : new_files) {
                std::string processed_file = "C:\\Users\\booki\\Desktop\\FunMoney\\DNA Storage\\concise_cpp\\mock_data\\" + file;
                std::cout << "New files added: " << processed_file << std::endl;
                std::string reads_trimmed_path = "C:\\Users\\booki\\Desktop\\FunMoney\\DNA Storage\\concise_cpp\\processed_data\\reads_trimmed_G";
                int lib_length = 100;  // Assume lib_length is available
                std::string design_file_path = "C:\\Users\\booki\\Desktop\\FunMoney\\DNA Storage\\concise_cpp\\deep_design.csv";  // Assume design_file_path is available

                read_preprocessor(processed_file, reads_trimmed_path, lib_length);
                binner(design_file_path, reads_trimmed_path);
            }
        }

        // Check for deleted files
        std::vector<std::string> deleted_files;
        for (const auto& [file, _] : previous_files) {
            if (current_files.find(file) == current_files.end()) {
                deleted_files.push_back(file);
            }
        }
        if (!deleted_files.empty()) {
            std::cout << "Files deleted: ";
            for (const auto& file : deleted_files) {
                std::cout << file << " ";
            }
            std::cout << std::endl;
        }

        // Check for modified files
        std::vector<std::string> modified_files;
        for (const auto& [file, timestamp] : current_files) {
            if (previous_files.find(file) != previous_files.end() && previous_files[file] != timestamp) {
                modified_files.push_back(file);
            }
        }
        if (!modified_files.empty()) {
            std::cout << "Files modified: ";
            for (const auto& file : modified_files) {
                std::cout << file << " ";
            }
            std::cout << std::endl;
        }

        // Update previous_files for the next iteration
        previous_files = current_files;
    }
}

int main() {
    std::string directory_to_monitor = "C:\\Users\\booki\\Desktop\\FunMoney\\DNA Storage\\concise_cpp\\mock_data";
    monitor_directory(directory_to_monitor, 5);
    return 0;
}

File: onChangeMonitor.py
import os
import time
from datetime import datetime
from binning import binner
from multi_threaded_preprocessor import read_preprocessor


# Function to scan the directory and return a dictionary of files with their modification times
def scan_directory(directory):
    files = {}
    for filename in os.listdir(directory):
        filepath = os.path.join(directory, filename)
        if os.path.isfile(filepath):
            files[filename] = os.path.getmtime(filepath)  # Get the modification time
    return files

# this parameter presents the lib length of 140
lib_length = 140

# path to the sequencing output - can be .txt or a .fastq file or a path to a folder with many .fastq files
# this is the file the includes all the basecalled reads that we want to processed
reads_path = "fastq_runid.fastq"

design_file_path = "./deep_design.csv"

# Function to monitor the directory for changes
def monitor_directory(directory, interval=5):
    print(f"Monitoring directory: {directory}")

    previous_files = scan_directory(directory)  # Initial snapshot of the directory
    # path (including file name) in which the trimmed reads will be stored
    while True:
        time.sleep(interval)  # Wait for the specified interval before scanning again

        current_files = scan_directory(directory)  # Scan the directory again

        # Check for new files
        new_files = [f for f in current_files if f not in previous_files]
        if new_files:
            for file in new_files:
                file = f'mock_data/{file}'
                print(f"New files added: {file}")
                reads_trimmed_path = "processed_data/reads_trimmed_" + datetime.today().strftime('%Y_%m_%d') + ".txt"
                read_preprocessor(file, reads_trimmed_path, lib_length)
                binner(design_file_path=design_file_path, reads_file_path=reads_trimmed_path)

        # Check for deleted files
        deleted_files = [f for f in previous_files if f not in current_files]
        if deleted_files:
            print(f"Files deleted: {deleted_files}")

        # Check for modified files
        modified_files = [f for f in current_files if f in previous_files and current_files[f] != previous_files[f]]
        if modified_files:
            print(f"Files modified: {modified_files}")

        # Update previous_files for the next iteration
        previous_files = current_files


# Specify the directory to monitor
directory_to_monitor = 'mock_data'

# Start monitoring
monitor_directory(directory_to_monitor, interval=5)  # Check every 5 seconds
